### PR TITLE
`std::panic::Location` is a lang_item, add `core::intrinsics::caller_location` (RFC 2091 3/N)

### PR DESCRIPTION
**Pull request by anp**
[Tracking issue](https://github.com/rust-lang/rust/issues/47809)
[RFC text](https://github.com/rust-lang/rfcs/blob/master/text/2091-inline-semantic.md)

@eddyb suggested doing this intrinsic implementation ahead of actually implementing the `#[track_caller]` attribute so that there's an easily tested intermediate step between adding the shim and wiring up the attribute.

### CODE REVIEW

File: src/libcore/panicking_location_intrinsic.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+use crate::fmt;
+use crate::panic::{Location, PanicInfo};
+
+#[cold]
+// never inline unless panic_immediate_abort to avoid code
+// bloat at the call sites as much as possible
+#[cfg_attr(not(feature="panic_immediate_abort"),inline(never))]
+#[lang = "panic"]
+pub fn panic(expr: &'static str, location: &Location<'_>) -> ! {
```
Comment (Centril) [inline src/libcore/panicking_location_intrinsic.rs]: Could we use conditional compilation at the function / statement level rather than the module level? (Makes the diff cleaner imo and easier to adjust later when bumping bootstrap.)

File: src/libcore/macros.rs [hunk 2]
```diff
@@ -1,6 +1,32 @@
...
+#[macro_export]
+#[allow_internal_unstable(core_panic)]
+#[stable(feature = "core", since = "1.6.0")]
+macro_rules! panic {
+    () => (
+        $crate::panic!("explicit panic")
+    );
+    ($msg:expr) => ({
+        #[allow(unused_unsafe)] // FIXME(anp) remove when `Location::caller()` impl'd
+        $crate::panicking::panic($msg, unsafe { $crate::intrinsics::caller_location() })
```
Comment (Centril) [inline src/libcore/macros.rs]: This shouldn't require `unsafe { ... }`.

File: src/librustc_codegen_llvm/builder.rs [hunk 3]
```diff
@@ -1068,34 +1069,17 @@ impl StaticBuilderMethods for Builder<'a, 'll, 'tcx> {
...
-            with_msg_components = [msg, filename, line, col];
-            &with_msg_components as &[_]
-        } else {
-            without_msg_components = [filename, line, col];
-            &without_msg_components as &[_]
-        };
-
-        let struct_ = self.const_struct(&components, false);
-        self.static_addr_of(struct_, align, Some(kind))
+        self.static_addr_of(struct_, align, None)
```
Comment (eddyb) [inline src/librustc_codegen_llvm/builder.rs]: You can add a FIXME for moving this into miri (which is better equipped to do this correctly even if e.g. the order of the fields changes). cc @oli-obk

File: src/librustc_codegen_ssa/mir/block.rs [hunk 4]
```diff
@@ -548,28 +528,33 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
...
             let layout = bx.layout_of(ty);
             if layout.abi.is_uninhabited() {
                 let loc = bx.sess().source_map().lookup_char_pos(span.lo());
-                let filename = Symbol::intern(&loc.file.name.to_string());
-                let line = bx.const_u32(loc.line as u32);
-                let col = bx.const_u32(loc.col.to_usize() as u32 + 1);
+                let location = bx.static_panic_location(&loc);
 
                 let str = format!(
                     "Attempted to instantiate uninhabited type {}",
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/block.rs]: Drive-by style nit (for @RalfJung?): messages should start with lowercase and types should have backticks around them, i.e.:
```rust
"attempted to instantiate uninhabited type `{}`"
```

File: src/test/ui/rfc-2091-track-caller/caller-location-intrinsic.rs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+// run-pass
+
+#![feature(core_intrinsics)]
+fn main() {
+    let loc = unsafe { core::intrinsics::caller_location() };
```
Comment (Centril) [inline src/test/ui/rfc-2091-track-caller/caller-location-intrinsic.rs]: The intrinsic should be whitelisted as safe to call.

File: src/librustc_codegen_ssa/mir/block.rs [hunk 6]
```diff
@@ -613,6 +613,22 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
             ReturnDest::Nothing
         };
 
+        if intrinsic == Some("caller_location") {
+            if let Some((_, target)) = destination.as_ref() {
+                let loc = bx.sess().source_map().lookup_char_pos(span.lo());
+                let location = bx.static_panic_location(&loc, "caller_loc");
+
+                if let ReturnDest::IndirectOperand(tmp, _) = ret_dest {
+                    super::operand::OperandValue::Immediate(location).store(&mut bx, tmp);
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/block.rs]: `Immediate` itself is imported, you can use it directly to make this shorter.

File: src/librustc_codegen_llvm/builder.rs [hunk 7]
```diff
@@ -1068,6 +1069,20 @@ impl StaticBuilderMethods for Builder<'a, 'll, 'tcx> {
...
+        let filename = self.const_str(filename);
+        let line = self.const_u32(loc.line as u32);
+        let col = self.const_u32(loc.col.to_usize() as u32 + 1);
+        let struct_ = self.const_struct(&[filename.0, filename.1, line, col], false);
+
+        let align = self.tcx.data_layout.aggregate_align.abi
+            .max(self.tcx.data_layout.i32_align.abi)
+            .max(self.tcx.data_layout.pointer_align.abi);
+        // FIXME(eddyb) move this into miri, it can be correct if e.g. field order changes
+        self.static_addr_of(struct_, align, Some(kind))
```
Comment (eddyb) [inline src/librustc_codegen_llvm/builder.rs]: You can hardcode the name here ("kind" is inaccurate, it's really just a global name), or not use one at all.

**Comment by anp** [comment]
@eddyb ping for review/r+? I'm thinking that it might be easier to make the macro changes in a separate PR after I'm able to roll the bootstrap compiler to include this.

**Comment by Centril** [comment]
When this is getting in a state that is ready for merging, can you please squash the commits into one?

**Comment by anp** [comment]
Good call! Done.

**Comment by anp** [comment]
From out of band discussion with @eddyb, I'm going to remove `static_panic_msg` in this PR.

**Comment by anp** [comment]
There are still some test failures after removing `static_panic_msg` and (by necessity) moving the panicking infra to use `Location` instead of `(&str, u32, u32)`, pertaining to const-eval and tests which rely on the type signature of the unstable lang items. Working on a follow-up.

File: src/test/ui/consts/const-eval/const_caller_location.rs [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+#![feature(core_intrinsics)]
+
+const LOCATION: &core::panic::Location = core::intrinsics::caller_location();
+
+fn main() {
+    assert_eq!(LOCATION.file(), file!());
+    assert_eq!(LOCATION.line(), 5);
+    assert_eq!(LOCATION.column(), 42);
+}
```
Comment (lzutao) [inline src/test/ui/consts/const-eval/const_caller_location.rs]:
```suggestion
}

```

File: src/librustc_codegen_ssa/mir/block.rs [hunk 9]
```diff
@@ -554,22 +535,10 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
...
-                );
-                let msg_str = Symbol::intern(&str);
-                let msg_file_line_col = bx.static_panic_msg(
-                    Some(msg_str),
-                    filename,
-                    line,
-                    col,
-                    "panic_loc",
-                );
+                let msg_str = format!("attempted to instantiate uninhabited type `{}`", ty);
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/block.rs]: I don't think I wanted this change bundled with this PR, but either as a separate PR or done by someone else, sorry for not being clear enough.

Comment (anp) [reply src/librustc_codegen_ssa/mir/block.rs]: back it out?

**Comment by anp** [comment]
@eddyb ping for review

File: src/librustc_codegen_ssa/mir/block.rs [hunk 10]
```diff
@@ -4,6 +4,7 @@ use rustc::ty::{self, Ty, TypeFoldable, Instance};
 use rustc::ty::layout::{self, LayoutOf, HasTyCtxt, FnTypeExt};
 use rustc::mir::{self, Place, PlaceBase, Static, StaticKind};
 use rustc::mir::interpret::PanicInfo;
+use rustc_mir::const_eval::const_caller_location;
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/block.rs]: I think @oli-obk's suggestion was to make this a query, like `const_field` is.

File: src/librustc_codegen_ssa/mir/block.rs [hunk 11]
```diff
@@ -1172,6 +1155,23 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         }
     }
 
+    fn make_caller_location(
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/block.rs]: I'd call this `get_caller_location`, but this is just a nit.

File: src/libcore/macros.rs [hunk 12]
```diff
@@ -2,21 +2,35 @@
...
     () => (
         $crate::panic!("explicit panic")
     );
     ($msg:expr) => ({
-        $crate::panicking::panic(&($msg, $crate::file!(), $crate::line!(), $crate::column!()))
+        #[cfg(bootstrap)]
+        { $crate::panicking::panic(&($msg, $crate::file!(), $crate::line!(), $crate::column!())) }
+
+        #[cfg(not(bootstrap))]
+        {
```
Comment (eddyb) [inline src/libcore/macros.rs]: Placing `#[cfg]` inside the expanded code, as opposed to duplicating the macro, can lead to slower expansion of *all* `panic!()` invocations, *and* it would depend on `--cfg bootstrap` *at the use site* (wherever that may be, i.e. in any code compiled with nightly `rustc`), as opposed to while building `libcore`.

Supposedly @Centril asked for this change but I can't find the original comment.

Comment (Centril) [reply src/libcore/macros.rs]: https://github.com/rust-lang/rust/pull/65664#discussion_r337051372

Comment (anp) [reply src/libcore/macros.rs]: Ah, I may have misunderstood the request about panicking.rs to also apply to the macros file. Let me know what I should do!

File: src/librustc_codegen_ssa/Cargo.toml [hunk 13]
```diff
@@ -31,4 +31,5 @@ rustc_errors = { path = "../librustc_errors" }
 rustc_fs_util = { path = "../librustc_fs_util" }
 rustc_incremental = { path = "../librustc_incremental" }
 rustc_index = { path = "../librustc_index" }
+rustc_mir = { path = "../librustc_mir" }
```
Comment (eddyb) [inline src/librustc_codegen_ssa/Cargo.toml]: Note that if you make `const_caller_location` a query then this disappears (this is a dependency we really don't want to add because it serializes two relatively large crates).

File: src/libcore/macros.rs [hunk 14]
```diff
@@ -20,6 +21,30 @@ macro_rules! panic {
...
+    );
+    ($msg:expr) => ({
+        let loc = $crate::panic::Location::internal_constructor(file!(), line!(), column!());
+        $crate::panicking::panic($msg, &loc)
+    });
+    ($msg:expr,) => (
+        $crate::panic!($msg)
+    );
+    ($fmt:expr, $($arg:tt)+) => ({
+        let loc = $crate::panic::Location::internal_constructor(file!(), line!(), column!());
```
Comment (eddyb) [inline src/libcore/macros.rs]: All of these invocations should be using `$crate::` just like the other version of the macro (otherwise users can shadow these uses with their own macros).

Comment (eddyb) [inline src/libcore/macros.rs]: One suggestion I had about this location was to use:
```rust
const LOC: &$crate::panic::Location = &$crate::panic::Location::internal_constructor(...);
```
to avoid codegen-ing a runtime call to the constructor and having a `Location` on a stack, effectively ensuring that codegen will be identical to before this PR and not cause any second-order effects etc.

File: src/librustc_mir/const_eval.rs [hunk 15]
```diff
@@ -504,13 +504,9 @@ pub fn const_field<'tcx>(
...
 
-pub fn const_caller_location<'tcx>(
-    tcx: TyCtxt<'tcx>,
-    param_env: ty::ParamEnv<'tcx>,
-    span: Span,
-) -> &'tcx ty::Const<'tcx> {
+pub fn const_caller_location<'tcx>(tcx: TyCtxt<'tcx>, span: Span) -> &'tcx ty::Const<'tcx> {
     trace!("const_caller_location: {:?}", span);
-    let mut ecx = mk_eval_cx(tcx, DUMMY_SP, param_env);
+    let mut ecx = mk_eval_cx(tcx, DUMMY_SP, ty::ParamEnv::reveal_all());
```
Comment (eddyb) [inline src/librustc_mir/const_eval.rs]: Heh, the `DUMMY_SP` could be `span` instead.

File: src/librustc_codegen_ssa/mir/block.rs [hunk 16]
```diff
@@ -1160,8 +1159,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         bx: &mut Bx,
         span: Span,
     ) -> Bx::Value {
-        let location = const_caller_location(bx.tcx(), bx.param_env(), span);
-
+        let location = bx.tcx().const_caller_location(span);
         bx.scalar_to_backend(
             location.val.try_to_scalar().unwrap(),
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/block.rs]: Wait, is there no way to lower an entire `ty::Const` without needing this entire dance?

File: src/libcore/macros.rs [hunk 17]
```diff
@@ -20,6 +21,38 @@ macro_rules! panic {
...
+    () => (
+        $crate::panic!("explicit panic")
+    );
+    ($msg:expr) => ({
+        const LOC: $crate::panic::Location<'_> = $crate::panic::Location::internal_constructor(
+            $crate::file!(),
+            $crate::line!(),
+            $crate::column!(),
+        );
+        $crate::panicking::panic($msg, &LOC)
```
Comment (eddyb) [inline src/libcore/macros.rs]: Put the reference in the `const` itself, to avoid relying on promotion.

File: src/librustc_codegen_ssa/mir/block.rs [hunk 18]
```diff
@@ -1172,6 +1154,18 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         }
     }
 
+    fn get_caller_location(
+        &self,
+        bx: &mut Bx,
+        span: Span,
+    ) -> Bx::Value {
+        let location = bx.tcx().const_caller_location(span);
+        match OperandRef::from_const(bx, location).val {
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/block.rs]: `.val.immediate()` does all this for you.

File: src/librustc_codegen_ssa/mir/block.rs [hunk 19]
```diff
@@ -1172,6 +1154,18 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         }
     }
 
+    fn get_caller_location(
+        &self,
+        bx: &mut Bx,
+        span: Span,
+    ) -> Bx::Value {
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/block.rs]: Should maybe be `OperandRef` but it's not that important.

File: src/libcore/panic.rs [hunk 20]
```diff
@@ -45,11 +45,20 @@ impl<'a> PanicInfo<'a> {
...
+    pub fn internal_constructor(
+        message: Option<&'a fmt::Arguments<'a>>,
+        location: &'a Location<'a>,
+    ) -> Self {
         struct NoPayload;
-        PanicInfo { payload: &NoPayload, location, message }
+        PanicInfo {
+            message,
+            payload: &NoPayload,
+            location: Location {
```
Comment (eddyb) [inline src/libcore/panic.rs]: Leave a FIXME comment about potentially making `PanicInfo` hold a reference to `Location`.

File: src/librustc_codegen_llvm/common.rs [hunk 21]
```diff
@@ -237,6 +219,13 @@ impl ConstMethods<'tcx> for CodegenCx<'ll, 'tcx> {
         unsafe { llvm::LLVMConstReal(t, val) }
     }
 
+    fn const_str(&self, s: Symbol) -> (&'ll Value, &'ll Value) {
```
Comment (eddyb) [inline src/librustc_codegen_llvm/common.rs]: Any chance you could get rid of this by using `OperandRef::from_const` with a string literal `ty::Const`?

File: src/librustc_typeck/check/intrinsic.rs [hunk 22]
```diff
@@ -143,6 +143,18 @@ pub fn check_intrinsic_type(tcx: TyCtxt<'_>, it: &hir::ForeignItem) {
                  ], tcx.types.usize)
             }
             "rustc_peek" => (1, vec![param(0)], param(0)),
+            "caller_location" => (
+                0,
+                vec![],
+                tcx.mk_ref(
```
Comment (eddyb) [inline src/librustc_typeck/check/intrinsic.rs]: Is there still an `mk_imm_ref` by any chance?

File: src/librustc_mir/interpret/intrinsics/caller_location.rs [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+        let u32_size = Size::from_bits(32);
+
+        let loc_ty = self.tcx.type_of(self.tcx.require_lang_item(PanicLocationLangItem, None))
+            .subst(*self.tcx, self.tcx.mk_substs([self.tcx.lifetimes.re_static.into()].iter()));
+        let loc_layout = self.layout_of(loc_ty)?;
+
+        let file_alloc = self.tcx.allocate_bytes(filename.as_bytes());
+        let file_ptr = Pointer::new(file_alloc, Size::ZERO);
+        let file = Scalar::Ptr(self.tag_static_base_pointer(file_ptr));
+        let file_len = Scalar::from_uint(filename.len() as u128, ptr_size);
```
Comment (eddyb) [inline src/librustc_mir/interpret/intrinsics/caller_location.rs]: cc @oli-obk There isn't a nice way to write a string literal constant as a `&str`?

**Review by eddyb — APPROVED** [review]
r=me with or without addressing comments (which are mostly optional nits at this point)

File: src/libcore/panic.rs (hunk 20 above)
Comment (anp) [reply src/libcore/panic.rs]: If you think it should hold a reference, I can make that change now. I had it previously but your question about its behavior with unsafe code made me think to not change the type decl.

Comment (eddyb) [reply src/libcore/panic.rs]: This is why I want a FIXME comment (and maybe even an issue) - the way it is right now works, and we should maybe investigate later whether it would be better not to copy the `Location` out.

**Review by eddyb — APPROVED** [review]

**Comment by anp** [comment]
Rolled this branch back to where it was passing CI but before any nits were addressed. OK to land as-is once CI passes?

File: src/libcore/panic.rs (hunk 20 above)
Comment (anp) [reply src/libcore/panic.rs]: https://github.com/rust-lang/rust/issues/65856

**Comment by anp** [comment]
@eddyb @Centril r+ with passing ci?

**Review by eddyb — APPROVED** [review]

**Review by eddyb — APPROVED** [review]

File: src/librustc/query/mod.rs [hunk 24]
```diff
@@ -466,6 +466,12 @@ rustc_queries! {
             no_force
             desc { "extract field of const" }
         }
+
+        query const_caller_location(key: (syntax_pos::Symbol, u32, u32)) -> &'tcx ty::Const<'tcx> {
+            eval_always
+            no_force
+            desc { "get a &core::panic::Location referring to a span" }
```
Comment (bjorn3) [inline src/librustc/query/mod.rs]: Not a span, but a position.

**Comment by eddyb** [comment]
@nnethercote Don't think you can do that after the fact. I should've done it before, or at least set `rollup=never`, my bad. Did the rollup that landed this have a perf impact?

**Comment by eddyb** [comment]
Wait, this might work:

@rust-timer build 4728d66206c82c98aa4e1fad751e8aae7489c242

**Comment by nnethercote** [comment]
@eddyb: see https://github.com/rust-lang/rust/pull/65885#issuecomment-547204122

**Comment by eddyb** [comment]
I've confirmed the codegen changes in the last commit of this PR are the cause of the regression and I'm testing a potential fix that doesn't involve reverting any chances from this PR at all.

File: src/librustc_mir/interpret/intrinsics/caller_location.rs [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+        let file_out = self.mplace_field(location, 0)?;
+        let file_ptr_out = self.force_ptr(self.mplace_field(file_out, 0)?.ptr)?;
+        let file_len_out = self.force_ptr(self.mplace_field(file_out, 1)?.ptr)?;
+        let line_out = self.force_ptr(self.mplace_field(location, 1)?.ptr)?;
+        let col_out = self.force_ptr(self.mplace_field(location, 2)?.ptr)?;
+
+        let layout = &self.tcx.data_layout;
+        let alloc = self.memory.get_mut(file_ptr_out.alloc_id)?;
+
+        alloc.write_scalar(layout, file_ptr_out, file.into(), ptr_size)?;
```
Comment (RalfJung) [inline src/librustc_mir/interpret/intrinsics/caller_location.rs]: Using the low-level `alloc` method here should be justified. In this case I think the justification is that we just created the allocation so we know we do not need to do any bounds checks.

Also, there should be some high-level comments here explaining what this does. An entire file without any comment is a dead giveaway.^^

Comment (eddyb) [reply src/librustc_mir/interpret/intrinsics/caller_location.rs]: I'd say there are two main improvements to be done here:
* make it trivial to create a `ty::Const` of a string literal (codegen could also use this in a couple places)
* make a "const aggregate" abstraction or similar, which gives you a `&'static Foo { x, y, z, ... }` where all the fields are given as `ty::Const`

Note sure if said abstraction should be a query, but coupled with the `const_field` query they could be an useful combination for operations on constant data from outside miri.

File: src/librustc_mir/interpret/intrinsics/caller_location.rs (hunk 23 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/intrinsics/caller_location.rs]: I don't think so, we only have a convenience method for reading a `&str`
